### PR TITLE
Fix SGX implementations of read/write_vectored.

### PR DESCRIPTION
**Pull request by sfackler**

### CODE REVIEW

File: src/libstd/sys/sgx/net.rs [hunk 1]
```diff
@@ -103,24 +103,22 @@ impl TcpStream {
...
-    pub fn read_vectored(&self, buf: &mut [IoVecMut<'_>]) -> io::Result<usize> {
-        let buf = match buf.get_mut(0) {
-            Some(buf) => buf,
-            None => return Ok(0),
-        };
-        self.read(buf)
+    pub fn read_vectored(&self, bufs: &mut [IoVecMut<'_>]) -> io::Result<usize> {
+        match bufs.iter_mut().find(|b| !b.is_empty()) {
+            Some(buf) => self.read(buf),
+            None => Ok(0),
```
Comment (jethrogb) [inline src/libstd/sys/sgx/net.rs]: Is this correct? On at least Linux & Windows, `read(&mut [])` blocks until the fd is ready for reading.

Comment (sfackler) [reply src/libstd/sys/sgx/net.rs]: Sure, we can change the default behavior to fall back to an empty buffer read.

**Comment by alexcrichton** [comment]
I think I may be missing some context, but there's two changes here right? One is the fix to the sgx module which looks for the first nonempty buffer instead of always using the first. the second change is to always defer to `read` no  matter what, falling back to a synthetic empty buffer if nothing is passed in or if everything is empty. The first change makes sense to me, but I'm less clear on what motivated the second change?

(nothing against it of course, just wanted to explore rationale)

**Comment by jethrogb** [comment]
This PR used to fix only SGX. But now it just generalizes this logic which is needed in multiple places and uses it everywhere it's needed.

**Comment by sfackler** [comment]
Yep, that's correct. The first part just fixes SGX and unifies the default implementations, and the second makes a behavior change to the defaults.

We want to specify that read_vectored/write_vectored should behave as a read/write would be to a buffer that's the concatenation of the vectored input buffers. Previously though, if you didn't pass any nonempty buffers we'd just return Ok(0) directly rather than actually making a read/write call with an empty buffer. @jethrogb pointed out that empty reads/writes actually do have interesting behavior so I changed the defaults up to ensure we always call read/write from the default read_vectored/write_vectored implementation.

Hopefully that makes sense! I can split the changes up into 2 PRs if you'd like.

**Comment by alexcrichton** [comment]
Ok thanks for confirming. I don't really understand where reading/writing an empty buffer would be significant, but so long as it's intentional sounds good to me

@bors: r+

**Pull request merged**